File: src/fragment/EntityTypeFragment.ts
import { Text } from "pixi.js";
import Fragment, { FragmentType, register } from "./Fragment";
import { StructEndecBuilder, PrimitiveEndecs } from 'KEndec';
import { Identifier } from "~/util";






export default class EntityTypeFragment extends Fragment {
    entityType: string;

    constructor(entityType: string) {
        super();

        this.entityType = entityType;
    }

    override toString(): string {
        return this.entityType;
    }

    override type(): FragmentType<EntityTypeFragment> {
        return ENTITY_TYPE;
    }
}

const ENTITY_TYPE = register("entity_type", 0x8877bb, 
    StructEndecBuilder.of1(
        Identifier.ENDEC
            .xmap((ident) => ident.toString(), Identifier.of)
            .fieldOf("entityType", (fragment: EntityTypeFragment) => fragment.entityType),
        (entityType) => new EntityTypeFragment(entityType)
    )
);

File: site/src/jsMain/kotlin/me/maplesyrum/tricksterStudio/spell/fragment/EntityTypeFragment.kt
package me.maplesyrum.tricksterStudio.spell.fragment

import me.maplesyrum.tricksterStudio.Identifier
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder


class EntityTypeFragment(val entityType: String) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.ENTITY_TYPE
    }

    override fun toString(): String {
        return this.entityType;
    }

    companion object {
        val ENDEC: StructEndec<EntityTypeFragment> = StructEndecBuilder.of(
            Identifier.ENDEC
                .xmap({ id -> id.toString() }, { str -> Identifier.of(str) })
                .fieldOf("entityType", EntityTypeFragment::entityType)
        ) { entityType: String -> EntityTypeFragment(entityType) }
    }
}
